bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/scriptengine/DelegatingScriptEngineWithInvocableAndAutocloseableAndSynchronization.java

```diff
@@ -26,47 +26,85 @@
 import org.eclipse.jdt.annotation.Nullable;
 
 /**
- * {@link ScriptEngine} implementation that delegates to a supplied ScriptEngine instance. Allows overriding specific
- * methods.
+ * {@link ScriptEngine} implementation that delegates to a supplied ScriptEngine instance and provides multi-thread
+ * synchronization. Allows overriding specific methods.
  *
  * @author Jonathan Gilbert - Initial contribution
  */
-public abstract class DelegatingScriptEngineWithInvocableAndAutocloseable<T extends ScriptEngine & Invocable & AutoCloseable>
+public abstract class DelegatingScriptEngineWithInvocableAndAutocloseableAndSynchronization<T extends ScriptEngine & Invocable & AutoCloseable>
         implements ScriptEngine, Invocable, AutoCloseable {
     protected T delegate;
+    /** Shared lock object for synchronization of multi-thread access */
+    protected final Object lock = new Object();
 
-    public DelegatingScriptEngineWithInvocableAndAutocloseable(T delegate) {
+    public DelegatingScriptEngineWithInvocableAndAutocloseableAndSynchronization(T delegate) {
         this.delegate = delegate;
     }
 
     @Override
     public @Nullable Object eval(String s, ScriptContext scriptContext) throws ScriptException {
-        return Objects.nonNull(delegate) ? delegate.eval(s, scriptContext) : null;
+        if (Objects.nonNull(delegate)) {
+            synchronized (lock) {
+                return delegate.eval(s, scriptContext);
+            }
+        } else {
+            return null;
+        }
```

Comment:
In general using `== null` or `!= null` is preferred over `Objects.isNull` and `Objects.nonNull`. The static methods from `Objects` should only be used when predicates in filter expressions are needed.

That aside, this type of null-check is not thread-safe because the value could be changed by another thread. Use a local copy instead:

```suggestion
        T localDelegate = delegate)
        if (localDelegate != null) {
            synchronized (lock) {
                return localDelegate.eval(s, scriptContext);
            }
        } else {
            return null;
        }
```

Otherwise LGTM, but I'm not an expert on the internals of this script engine either.

Reply:
While thinking about your comment, I am wondering if we need those null checks or if we should annotate `delegate` with `@NonNull`. It is not initialized in this class, but it is in the constructor of `OpenhabGraalJSScriptEngine`. So WDYT?